src/OrchardCore.Modules/OrchardCore.Media/Shortcodes/ShareShortcodeProvider.cs

```diff
@@ -0,0 +1,174 @@
+        private static readonly HashSet<string> Shortcodes = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        {
+            "share",
+            "doc",
+            "download"
+        };
+
+        private readonly IMediaFileStore _mediaFileStore;
+        private readonly IHtmlSanitizerService _htmlSanitizerService;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly ResourceManagementOptions _options;
+
+        public ShareShortCodeProvider(
+            IMediaFileStore mediaFileStore,
+            IHtmlSanitizerService htmlSanitizerService,
+            IHttpContextAccessor httpContextAccessor,
+            IOptions<ResourceManagementOptions> options)
+        {
+            _mediaFileStore = mediaFileStore;
+            _htmlSanitizerService = htmlSanitizerService;
+            _httpContextAccessor = httpContextAccessor;
+            _options = options.Value;
+        }
+
+        #region https://stackoverflow.com/a/40361205/549306
+        private static readonly Uri SomeBaseUri = new Uri("http://canbeanything");
+
+        private static string GetFileNameFromUrl(string url)
+        {
+            Uri uri;
+            if (!Uri.TryCreate(url, UriKind.Absolute, out uri))
+            {
+                uri = new Uri(SomeBaseUri, url);
+            }
+
+            return Path.GetFileName(uri.LocalPath);
+        }
+
+        #endregion
+
+        public ValueTask<string> EvaluateAsync(string identifier, Arguments arguments, string content, Context context)
+        {
+            // we can't process what we don't understand, and we are discarding or filtering out anything we don't understand to keep things safe
+            if (!Shortcodes.Contains(identifier))
+            {
+                return Null;
+            }
+
+            // critical that at least the content or the href/url/source is set
+            if (String.IsNullOrWhiteSpace(content) && !IsSet(arguments, ShortcodeProperty.Url))
+            {
+                return ShareShortcode;
+            }
+
+            string url;
+            string href;
+            // sort out and infer as necessary the description (content) and resource location (url and href)
+            if (IsSet(arguments, ShortcodeProperty.Url))
+            {
+                url = UrlFixer(arguments.NamedOrDefault(ShortcodeProperty.Url.Name));
+                if (String.IsNullOrWhiteSpace(content))
+                {
+                    content = GetFileNameFromUrl(url).Split(FileExtensionSeperator)[0];
+                }
+            }
+            else
+            {
+                url = UrlFixer(content);
+                // because both the content and the href can't be empty, we presume the content was the href in this branch, and replace a shortened, friendlier name for the resulting content
+                content = GetFileNameFromUrl(content).Split(FileExtensionSeperator)[0];
```

Comment:
I would have used the original url as the default content if it's not provided. Like markdown is doing.

Reply:
I have made the suggested change and it will be in the resulting submission.

The caveat, tenancy and media-store prefixes change the URL, and that is reflected back into the inferred content.